src/dotty/tools/dotc/core/unpickleScala2/Scala2Unpickler.scala

```diff
@@ -147,6 +147,51 @@ object Scala2Unpickler {
     denot.info = ClassInfo( // final info
       denot.owner.thisType, denot.classSymbol, parentRefs, declsInRightOrder, ost)
   }
+
+  /** Adapt arguments to type parameters so that variance of type lambda arguments
+   *  agrees with variance of corresponding higherkinded type parameters. Example:
+   *
+   *     class Companion[+CC[X]]
+   *     Companion[List]
+   *
+   *  with adaptArgs, this will expand to
+   *
+   *     Companion[[X] => List[X]]
+   *
+   *  instead of
+   *
+   *      Companion[[+X] => List[X]]
+   *
+   *  even though `List` is covariant. This adaptation is necessary to ignore conflicting
+   *  variances in overriding members that have types of hk-type parameters such as `Companion[GenTraversable]`
+   *  or `Companion[ListBuffer]`. Without the adaptation we would end up with
+   *
+   *      Companion[[+X] => GenTraversable[X]]
+   *      Companion[[X] => List[X]]
+   *
+   *  and the second is not a subtype of the first. So if we have overridding memebrs of the two
+   *  types we get an error.
+   */
+  def adaptArgs(tparams: List[Symbol], args: List[Type])(implicit ctx: Context): List[Type] = tparams match {
+    case tparam :: tparams1 =>
+      val boundLambda = tparam.infoOrCompleter match {
+        case TypeBounds(_, hi) => hi.LambdaClass(forcing = false)
+        case _ => NoSymbol
+      }
+      def adaptArg(arg: Type): Type = arg match {
+        case arg: TypeRef if arg.symbol.isLambdaTrait =>
```

Comment:
Unrelated to this change, but the definition of `isLambdaTrait`:

```
    final def isLambdaTrait(implicit ctx: Context): Boolean =
      isClass && name.startsWith(tpnme.LambdaPrefix)
```

Looks prone to falsely matching:

```
scala> class `Lambda|`
defined class Lambda$bar
```